### ISSUE TITLE
Looking for guidance

### ISSUE THREAD

**Issue by JaneSoo** [ibus/ibus-m17n#65]
Hello, I am trying to find a way to fix the Khmer(km- Cambodian) input keyboard for Linux/Fedora. I ended up here after searching for a while. Could anyone tell me if this is the right place? If yes, is it possible to get help with getting started? (sorry for creating an issue, I don't know where else to reach out). Thanks in advance!

**Comment by mike-fabian** [comment]
It is possible that you are at the right place here.

But can you explain what the problem is?

Are you already using ibus-m17n to input Khmer?

Currently there is only one Khmer input method in m17n-db, i.e. only one Khmer input method which can be used with ibus-m17n (or ibus-typing-booster). This one:

```
mfabian@hathi:/local/mfabian/src/ibus-m17n (no-hide-before-commit *)
$ grep -i khmer /usr/share/m17n/*.mim
/usr/share/m17n/km-yannis.mim:;; km-yannis.mim -- Khmer input method suggested by Dr. Yannis Haralambous
/usr/share/m17n/km-yannis.mim:(description "Khmer input method suggested by Dr. Yannis Haralambous.
mfabian@hathi:/local/mfabian/src/ibus-m17n (no-hide-before-commit *)
$ ibus list-engine | grep yannis
  m17n:km:yannis - km-yannis (m17n)
mfabian@hathi:/local/mfabian/src/ibus-m17n (no-hide-before-commit *)
$ 
```

Are you using this km-yannis Khmer input method? Is there anything wrong with it?
Or do you have problems setting up ibus to use it?

**Comment by JaneSoo** [comment]
Hi,
I am on Fedora 38 and I didn't do any additional configuration for input method. I ran the two commands as you did above and got the same results as yours.

Yes, there are problems with the input method. So to construct a Khmer word, there are many possible combinations. For example, consonant + vowel ក​+ំ => កំ​ this one works but something like this ក +​្+ត should be ក្តា​  doesn't work on terminal. I guess nobody would use it on a terminal but I just want to get started to improve it. Let me know if you have any tips

**Comment by mike-fabian** [comment]
> Hi, I am on Fedora 38 and I didn't do any additional configuration for input method. I ran the two commands as you did above and got the same results as yours.

Yes, but that just shows that ibus, m17n-db, and the km-yannis input method are installed.

I am still not sure what you are actually using.

Are you using the Gnome desktop?

If yes, what do you see when you open the gnome-control-center, Keyboard section?

Here is a screenshot from a Virtual machine running Fedora 39 and the Gnome Wayland desktop (You use Fedora 38, but that does not matter here):

![Screenshot](https://github.com/ibus/ibus-m17n/assets/2330175/5b209d2f-992f-474d-9f5b-01d8fdcfe81e)

You see that I have two Khmer related entries there:

- Khmer (km-yannis (m17n))
- Khmer (Cambodia)

The first one is the km-yannis input method of m17n-db used via ibus-m17n

The second is the Xorg keyboard layout from 

```
$ head /usr/share/X11/xkb/symbols/kh 
// Khmer Unicode keyboard layout for XFree86 and XOrg,
// by the National Information Communications Technology Development Authority
// (NiDa) of Cambodia -- http://www.nida.gov.kh/.
//
// version:  1.0.1
// date:  14.09.2005
// author:  Jens Herden  (jens at khmeros.info)

default partial alphanumeric_keys
xkb_symbols "basic" {
```
Which one are you using?

If you are using km-yannis, you should see a "ក" at the top in the Gnome panel.

If you are using the Khmer (Cambodia) keyboard layout, yhou should see "km" at the top in the Gnome panel.

**Comment by mike-fabian** [comment]
> For example, consonant + vowel ក​+ំ => កំ​ this one works but something like this ក +​្+ត should be ក្តា​ doesn't work on terminal.

ក U+1780 KHMER LETTER KA
ំ U+17C6 KHMER SIGN NIKAHIT
កំ U+1780 KHMER LETTER KA U+17C6 KHMER SIGN NIKAHIT
ក U+1780 KHMER LETTER KA
្ U+17D2 KHMER SIGN COENG
ត U+178F KHMER LETTER TA
ក្ត U+1780 KHMER LETTER KA U+17D2 KHMER SIGN COENG U+178F KHMER LETTER TA
ក្តា U+1780 KHMER LETTER KA U+17D2 KHMER SIGN COENG U+178F KHMER LETTER TA U+17B6 KHMER VOWEL SIGN AA

​Screenshot of this in xfce4-terminal:

![Screenshot](https://github.com/ibus/ibus-m17n/assets/2330175/7e900693-fac2-4cdd-883b-a1f039cc564f)

**Comment by mike-fabian** [comment]
Screenshot of this in vim in gnome-terminal:

![Screenshot](https://github.com/ibus/ibus-m17n/assets/2330175/fbe18042-d1cb-461e-990f-8139c70324c7)

**Comment by mike-fabian** [comment]
Screenshot of this in gedit:

![Screenshot](https://github.com/ibus/ibus-m17n/assets/2330175/089d3650-da36-44a0-b548-03e17a127ede)

**Comment by mike-fabian** [comment]
I think that the terminals don’t support Khmer well is a known problem.

But in gedit it looks correct, doesn’t it?

**Comment by mike-fabian** [comment]
When I look at how the text quoted from you is displayed in vim in the terminals, I notice several <200b>:

```
> For example, consonant + vowel ក<200b>+ំ => កំ<200b> this one works but something like this ក +<200b>+ត should be ក្តា<200b> doesn't work on terminal.
```

U+200B ZERO WIDTH SPACE

That would be quite hard to enter using the km-yannis input method, there is no such character in km-yannis, it would be hard to input that ZERO WIDTH SPACE using km-yannis.

**But**, that character can be easily typed using the Xorg keyboard layout:

```
mfabian@hathi:~
$ grep -i 200b  /usr/share/X11/xkb/symbols/kh 
    key <SPCE>	{[ U200b,	space,		nobreakspace,	none	]};
mfabian@hathi:~
$ 
```

Just hitting the space bar will input U+200B if you are using that layout. To get a “normal” space you need to type Shift+space. 

That makes me guess that you are probably using the Xorg keyboard layout, not the m17n input method km-yannis.

**Comment by JaneSoo** [comment]
Hi, thank you so much for detailed explanation and test. 
I was using khmer(Cambodian) not the km-yannis. It works well everywhere else except for terminal.
Here is the example: វាធ្វើការបានត្រឹមត្រូវ

But now I tried with km-yannis, I found a problem
វaមiនតញរwមតញរUវ

I typed them the same way but one works and the other one doesn't work. Unless there are different way to construct work specifically for km-yannis. 


And yes I use Gnome desktop. By Xorg keyboard layout you meant the normal khmer (Cambodian)? sorry I am new to the input stuff.

**Comment by mike-fabian** [comment]
> Hi, thank you so much for detailed explanation and test. I was using khmer(Cambodian) not the km-yannis. It works well everywhere else except for terminal. Here is the example: វាធ្វើការបានត្រឹមត្រូវ

Yes, that it does not work in the terminal is a bugs and not implemented features in the terminals. That doesn't have anything to do with how you type it.

Whether you type it or paste it or save some Khmer text to a file khmer.txt (for example with some program like gedit where it works) and then do

```
cat khmer.txt
```

in a terminal, the terminal will not display the text correctly either. So you could try opening bugs about the terminals not being able to display Khmer text correctly, but that will not be easy to fix.

> But now I tried with km-yannis, I found a problem វaមiនតញរwមតញរUវ
> 
> I typed them the same way but one works and the other one doesn't work. Unless there are different ways to construct the word specifically for km-yannis.
> 
> And yes I use Gnome desktop. By Xorg keyboard layout you meant the normal khmer (Cambodian)? sorry, I am new to the input stuff.

Yes, “Khmer (Cambodian)” is the Xorg keyboard layout which is defined in the file `/usr/share/X11/xkb/symbols/kh`. You can look into that file with an editor or with `less /usr/share/X11/xkb/symbols/kh`. The file starts like this:

```
// Khmer Unicode keyboard layout for XFree86 and XOrg,
// by the National Information Communications Technology Development Authority
// (NiDa) of Cambodia -- http://www.nida.gov.kh/.
//
// version:  1.0.1
// date:  14.09.2005
// author:  Jens Herden  (jens at khmeros.info)

default partial alphanumeric_keys
xkb_symbols "basic" {

    name[Group1]= "Khmer (Cambodia)";

//  Four levels:   normal       Shift           right Alt       Shift+right Alt

    key <TLDE>  {[guillemotleft,guillemotright, U200d,          none    ]};
    key <AE01>  {[ U17e1,       exclam,         U200c,          U17f1   ]};
... more stuff like this ...
```

So for each key it defines 4 characters, first column is the character you get pressing only that key, second column is what you get when you press that key with Shift, third column is what you get when you press that key together with the right Alt key, fourth column is what you get when you press that key with Shift **and** the right Alt key.

`<AE01>` is the key which has the label 1 on an US English layout:

```
$ grep '<AE01>' /usr/share/X11/xkb/symbols/us  | head -1 
    key <AE01>	{[	 1,	 exclam		]};

```

Using such keyboard layouts, each key can only output **exactly** one character, not more.

But for some keys, one wants to get a result which is more than just one character. To work around the "one character limitation" of these keyboard layouts, compose is used.
The compose file contains:

```
mfabian@hathi:~
$ grep -i khmer /usr/share/X11/locale/en_US.UTF-8/Compose 
# Khmer digraphs
<U17ff>	: "ាំ"	# KHMER VOWEL SIGN AA plus KHMER SIGN NIKAHIT
<U17fe>	: "ោះ"	# KHMER VOWEL SIGN OO plus KHMER SIGN REAHMUK
<U17fd>	: "េះ"	# KHMER VOWEL SIGN E plus KHMER SIGN REAHMUK
<U17fc>	: "ុំ"	# KHMER VOWEL SIGN U plus KHMER SIGN NIKAHIT
<U17fb>	: "ុះ"	# KHMER VOWEL SIGN U plus KHMER SIGN REAHMUK
<dead_currency> <I>			: "៛"	U17DB # KHMER CURRENCY SYMBOL RIEL
mfabian@hathi:~
$ 
```

U+17FB, U+17FC, U+17FD, U+17FE, and U+17FF are in the Khmer code page of Unicode, but they are all unassigned.

See: https://en.wikipedia.org/wiki/Khmer_(Unicode_block)

These unassigned characters can be produced by the Khmer keyboard layout:

```
mfabian@hathi:~
$ grep -E '17f(b|c|d|e|f)'  /usr/share/X11/xkb/symbols/kh 
    key <AC01>	{[ U17b6,	U17ff,		none,		U19ec	]};
    key <AC10>	{[ U17be,	U17fe,		U17d6,		U19f5	]};
    key <AB04>	{[ U179c,	U17fd,		none,		U19fa	]};
    key <AB08>	{[ U17fc,	U17fb,		comma,		U19fe	]};
mfabian@hathi:~
$ 
```

For example, the `<AC01>` key has U17ff when you press that key together with Shift.
`<AC01>` is the key which is labelled `a` on the US English layout:

```
mfabian@hathi:~
$ grep '<AC01>' /usr/share/X11/xkb/symbols/us  | head -1 
    key <AC01>	{[	 a,	 A		]};
mfabian@hathi:~
$
```

So if you press that key with shift, it first produces U+17FF, but that is not what you really want, if compose is working correctly, then it should be replaced by

"ាំ"	# KHMER VOWEL SIGN AA plus KHMER SIGN NIKAHIT

Until very recently, these single key compose sequences (which are used in Khmer and Arabic) were broken, but they have recently been fixed in ibus (and also in ibus-typing-booster).

On you Fedora 38 system, you should have ibus >= 1.5.28, and this replacement using compose should work for you as long as ibus is running and in use.

These single key compose sequences are still broken in gtk, i.e. if you avoid using ibus by doing for example:

```
GTK_IM_MODULE=gtk-im-context-simple gedit
GTK_IM_MODULE=gtk-im-context-simple gnome-text-edit
```

Then it does not work. (Using the environment variable `GTK_IM_MODULE=gtk-im-context-simple` specifies that the compose support in Gtk instead of ibus should be used and in gtk this still does not work).

**Comment by mike-fabian** [comment]
> But now I tried with km-yannis, I found a problem វaមiនតញរwមតញរUវ
> 
> I typed them the same way but one works and the other one doesn't work. Unless there are different ways to construct the word specifically for km-yannis.

km-yannis works completely different from the Khmer (Cambodia) keyboard layout.

To find out how it works, you can read the file `/usr/share/m17n/km-yannis.mim`.

In an Xorg keyboard layout, each key produces exactly one characters (except for the compose hack expanding such a single character into something else which I explained in my last comment).

km-yannis.mim is a m17n input method which does not have this limitation. A key pressed could output anything, it could output any sequence of characters, it does not have to be a single character. Also, a sequence of input keys pressed in a certain order may output something.

The km-yannis file is pretty short and looks like this:

https://git.savannah.nongnu.org/cgit/m17n/m17n-db.git/tree/MIM/km-yannis.mim

For example it contains:

```
(map
 (consonant
  ("k" "ក")
  ("kh" "ខ")
  ("K" "គ")
  ("Kh" "ឃ")
  ("KH" "ឃ")
  ...
```

That is pretty easy to read, typing `k` produces "ក", typing `kh` produces "ខ", typing `K` produces "គ" etc.

Below are other sections like

```
 (wirama
  ("/" "្"))

 (vowel
  ("A" "ា")
  ("e" "ិ")
  ...
```

And at the bottom there is something like

```
(state
 (init
  (consonant (shift c-state))
  (vowel)
  (quality)
  (diacritic)
  (ind-vowel)
  (misc))

 (c-state
  (wirama (shift cw-state))
  (space (shift init)))

 (cw-state
  (consonant (shift cwc-state))
  (space (shift init)))

 (cwc-state
  (wirama (shift cwcw-state))
  (space (shift init)))

 (cwcw-state
  (consonant (shift init))))
```

So in the beginning it satrts in the `init` state and you can type something from the consonant, vowel, quality, diacritic, ind-vowel, or misc maps.

If you type a consonant, it switches to the `c-state` where you can either type something from the wirama map:

```
 (wirama
  ("/" "្"))
```

I.e. you can type `/` to add "្" to the consonant and shift to the cw-state or type a space to add nothing and switch back to the init state. And so on, there are a few more states but you can probably guess how that works now.

**Comment by JaneSoo** [comment]
Wow you just taught me a lot of things. Thank you so much. I'm still processing information though.

>>https://git.savannah.nongnu.org/cgit/m17n/m17n-db.git/tree/MIM/km-yannis.mim
This is really interesting, someone who didn't know about it like me would think that km-yannis didn't work but it actually has a completely different way of constructing a word.

I am playing around with both of the inputs (khmer-unicode from Nida and km-yannis). I am trying them on the normal text editor (not on the terminal)
It seems that after switching between the two keyboard layouts, the input system got confused and I could type correctly in km-yannis IN A WAY as I type in khmer-unicode. Meaning that if I want to write ខ​ in km-yannis, I only just type `x` instead of `kh`. (But anyway, I believe it's just a glitch

**Comment by mike-fabian** [comment]
> Meaning that if I want to write ខ​ in km-yannis, I only just type x instead of kh. (But anyway, I believe it's just a glitch)

Ah, no that is not a glitch. I’ll try to explain what happens:

If you use the Khmer (Cambodia) layout, typing `x` will give you ខ U+1781 KHMER LETTER KHA.

If you then switch to km-yannis and type `x`, the keyboard layout is not changed, it is still the Khmer (Cambodia) layout. So you still get ខ U+1781 KHMER LETTER KHA when typing `x`. And km-yannis has no rules what should happen when it gets ខ as input so it just passes it through unchanged.

km-yannis is designed to work with an US English keyboard layout (or any other keyboard layout which can produce the characters used as input. For example

```
("y" "យ")
```

needs the input of `y` to produce "យ". I could use an US English keyboard to type the `y`, it is to the right of the `t` key in the top row (US English has a QWERTY layout). I could also use a German layout to type the `y`, but as the German layout is QWERTZ:

https://en.wikipedia.org/wiki/German_keyboard_layout#/media/File:KB_Germany.svg

the `y` key is in a different position, it is in the bottom row to the left of the `x`.

But if you are using the Khmer (Cambodia) layout, you cannot produce the input km-yannis expects, so it does not work as intended.

So it matters which keyboard layout was used before switching to km-yannis.

**Comment by mike-fabian** [comment]
As having to remember which keyboard layout was used last before switching to km-yannis, one can select an option 

☑️ Use US keyboard layout 

in the setup tool of km-yannis. See this screenshot:

![Screenshot](https://github.com/ibus/ibus-m17n/assets/2330175/5c33bf04-7d82-461c-9ca8-afedc7ca5e65)

**Comment by mike-fabian** [comment]
Oh, there seems to be a bug, the option I described in my last comment does not seem to work. Checking ...

**Comment by JaneSoo** [comment]
What is the bug? It seems to work as you described.

Anyway, that's interesting how the input works. You said that km-yannis is designed to work with a US keyboard. So I assume Khmer(Cambodia) does a similar thing. So ខ in Khmer(Cambodia) is `x`, then when I switch to km-yannis, wouldn't it understand the input is `x` rather than `ខ`. Sorry if it's a dump question.

**Comment by mike-fabian** [comment]
I opened this new bug for the problem that the ibus-m17n option to enforce US keyboard layout does not work when coming from a layout which does not produce ASCII letters:
https://github.com/ibus/ibus-m17n/issues/66

**Comment by JaneSoo** [comment]
Thank you for sharing the bug issue

**Comment by JaneSoo** [comment]
> > What is the bug? It seems to work as you described.
> 
> No, unfortunately it doesn’t. I'll open a bug in a few minutes, maybe it is easier to understand if you look at that bug when I have opened it.
> 
> > Anyway, that's interesting how the input works. You said that km-yannis is designed to work with a US keyboard.
> 
> Yes, mostly although any other layout which can produce the needed Latin letters like `x` can be used as well.
> 
> > So I assume Khmer(Cambodia) does a similar thing.
> 
> No, Khmer (Cambodia) is a keyboard layout itself, not an input method. It does not get an `x` as an input and transforms it into something else (which is what km-yannis does) but it sees a press of the `<AB02>` key and then outputs U+1781 which is `ខ`. The same key `<AB02>` on the US layout happens to output an `x`. But when the Khmer layout is used, there is no `x`, only `ខ` and km-yannis cannot use `ខ` as input:
> 
> US Layout has `x` on the `<AB02>` key:
> 
> ```
> mfabian@hathi:~
> $ grep '<AB02>' /usr/share/X11/xkb/symbols/us  | head -1 
>     key <AB02>	{[	 x,	 X		]};
> ```
> 
> Khmer layout has U+1781 `ខ` on the `<AB02>` key:
> 
> ```
> mfabian@hathi:~
> $ grep '<AB02>' /usr/share/X11/xkb/symbols/kh  | head -1 
>     key <AB02>	{[ U1781,	U1783,		none,		U19f8	]};
> mfabian@hathi:~
> $ 
> ```
> 
> > So ខ in Khmer(Cambodia) is `x`, then when I switch to km-yannis, wouldn't it understand the input is `x` rather than `ខ`. Sorry if it's a dump question.
> 
> No, in that case the km-yannis gets an `ខ` as input and cannot use that and therefore just passes it through. And unfortunately the option to automatically switch to US layout when switching to km-yannis does not work (I’ll report a bug about that).
> 
> Not a dumb question, this is difficult to understand for new users.

Very well explained, now that makes sense. Let me know if I can help anything with the bug

**Comment by mike-fabian** [comment]
> Let me know if I can help anything with the bug

I fixed the bug https://github.com/ibus/ibus-m17n/issues/66 already. The fix is included in  the 1.4.23 release of ibus-m17n: https://github.com/ibus/ibus-m17n/releases/tag/1.4.23

But surely you can help improving Khmer input!

The wikipedia page https://en.wikipedia.org/wiki/Khmer_keyboard says:

The Nida Unicode Khmer keyboard is currently the most widely used in Cambodia

and the wikipedia page has this picture https://en.wikipedia.org/wiki/Khmer_keyboard#/media/File:Khmer_unicode_layout.svg

Does the Khmer (Cambodia) layout we have in Linux agree 100% with this, or are there some differences?

Is the km-yannis input method useful for you?

What other things could we do to improve Khmer input?

**Comment by mike-fabian** [comment]
Nothing happened in this issue for a while, I think I should better close it. Feel free to reopen or create a new issue if you still have problems or ideas.
